lib/wibox/layout/stack.lua

```diff
@@ -0,0 +1,166 @@
+-- @treturn boolean If the operation is successful
+-- @name replace_widget
+-- @class function
+
+--- Swap 2 widgets in a layout
+-- @tparam number index1 The first widget index
+-- @tparam number index2 The second widget index
+-- @treturn boolean If the operation is successful
+-- @name swap
+-- @class function
+
+--- Swap 2 widgets in a layout
+-- If widget1 is present multiple time, only the first instance is swapped
+-- @param widget1 The first widget
+-- @param widget2 The second widget
+-- @tparam[opt=false] boolean recursive Digg in all compatible layouts to find the widget.
+-- @treturn boolean If the operation is successful
+-- @name swap_widgets
+-- @class function
+
+--- Insert a new widget in the layout at position `index`
+-- @tparam number index The position
+-- @param widget The widget
+-- @treturn boolean If the operation is successful
+-- @name insert
+-- @class function
+
+--- Layout a stack layout. Each widget get drawn on top of each other
+-- @param context The context in which we are drawn.
+-- @param width The available width.
+-- @param height The available height.
+function stack:layout(context, width, height)
+    local result = {}
+    local spacing = self._spacing
+
+    for k, v in pairs(self.widgets) do
+        table.insert(result, base.place_widget_at(v, spacing, spacing, width - 2*spacing, height - 2*spacing))
+        if self._top_only then break end
+    end
+
+    return result
+end
+
+--- Get if only the first stack widget is drawn
+-- @return If the only the first stack widget is drawn
+function stack:get_display_top_only()
+    return self._top_only
+end
+
+--- Only draw the first widget of the stack, ignore others
+-- @tparam boolean top_only Only draw the top stack widget
+function stack:set_display_top_only(top_only)
+    self._top_only = top_only
+end
+
+--- Raise a widget at `index` to the top of the stack
+-- @tparam number index the widget index to raise
+function stack:raise(index)
+    if not index or index > #self.widgets or index < 1 then return end
+
+    if layout == self and idx ~= 1 then
+        self.widgets[idx], self.widgets[1] = self.widgets[1], self.widgets[idx]
+        self:emit_signal("widget::layout_changed")
+        self:emit_signal("widget::redraw_needed")
+    end
+end
+
+--- Raise the first instance of `widget`
+-- @param widget The widget to raise
+-- @tparam[opt=false] boolean recursive Also look deeper in the hierarchy to
+--   find the widget
+function stack:raise_widget(widget, recursive)
+    local idx, layout = self:index(widget, recursive)
+
+    if not idx or not layout then return end
+
+    -- Bubble up in the stack until the right index is found
+    while layout and layout ~= self do
+        idx, layout = self:index(layout, recursive)
+    end
+
+    if layout == self and idx ~= 1 then
+        self:raise(idx)
+    end
+end
+
+local function new(dir, widget1, ...)
+    local ret = fixed.horizontal(...)
```

Comment:
A bit of a hack, but ok...

You recently wrote something about awesome needing better docs. Could you add something for this? E.g. something somewhere should explain what the indices mean and that index 1 is the bottom-most widget (if I understand `:layout()` correctly etc.